sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala

```diff
@@ -292,7 +292,7 @@ class OptimizeJsonExprsSuite extends PlanTest with ExpressionEvalHelper {
 
     Seq("""{"a":1, "b":2, "c": 123, "d": "test"}""", null).foreach(v => {
       val row = create_row(v)
-      checkEvaluation(e1, e2.eval(row), row)
+      checkEvaluation(e1, replace(e2).eval(row), row)
```

Comment:
I missed this. `JsonToStructs` needs to carry the semantic information so that the optimizer rule can match it and do something. Replacing it with `Invoke` breaks it.

I propose to revert it and revisit other PRs that reimplement expressions with `RuntimeReplaceable`.

cc @panbingkun

Reply:
also cc @MaxGekk

Reply:
Which optimization rule or scenario is not effective?
I can investigate it.
Thanks!

Reply:
is this one? `OptimizeCsvJsonExprs`?
https://github.com/apache/spark/blob/e55511c98e14115dd023f9014372fc7f23d9ef6c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeCsvJsonExprs.scala#L39

Reply:
Or would it break this rule if we `manually` implement it instead of using `Invoke`?

Reply:
The problem is, it's not only about matching, but also about replacing. In `OptimizeCsvJsonExprs`, we need to replace the `JsonToStructs` with a new one using a different schema, and it's very hard to do so with matching `Invoke`.

In the long term, I think we should make `RuntimeReplaceable` really do replacement at runtime, instead of at the beginning of the optimizer, especially for the one with `StaticInvoke`/`Invoke`.

Reply:
Okay, let me revert it first and keep it implemented manually. 
Also, `CsvToStructs` may also have this issue, right? 
Let's manually implement its codegen?

Reply:
Yea let's manually implement codegen first.